File: visualize_strategy.py
import matplotlib
matplotlib.use('Agg')  # 使用非交互式后端，避免显示窗口
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
from room_inspection_greedy import RoomInspectionGreedy

def visualize_strategy(person1, person2, algorithm, save_path='room_inspection_strategy.png'):
    """
    可视化房间检查策略
    
    参数：
    - person1: 人员1对象
    - person2: 人员2对象
    - algorithm: RoomInspectionGreedy算法实例
    - save_path: 保存图片的路径
    """
    # 获取数据
    hallway_length = algorithm.hallway_length
    exit1_pos = 0
    exit2_pos = hallway_length
    corridor_y = 0
    
    # 创建图形
    fig, ax = plt.subplots(figsize=(14, 8))
    ax.set_aspect('equal')
    ax.grid(True, alpha=0.3, linestyle='--')
    
    # 不绘制走廊背景，只添加出口标签
    ax.text(exit1_pos - 1, corridor_y, 'Exit1', fontsize=12, 
            fontweight='bold', ha='right', va='center')
    ax.text(exit2_pos + 1, corridor_y, 'Exit2', fontsize=12, 
            fontweight='bold', ha='left', va='center')
    
    # 房间位置和样式
    room_offset = 2.5  # 房间距离走廊的偏移
    room_size = 100    # 房间标记大小
    
    # 绘制房间
    left_rooms = [room for room in algorithm.rooms if room.side == 'left']
    right_rooms = [room for room in algorithm.rooms if room.side == 'right']
    
    # 左侧房间（在走廊上方）
    for room in left_rooms:
        pos = room.distance_from_exit1
        # 绘制房间标记
        ax.scatter(pos, corridor_y + room_offset, s=room_size*2, 
                  c='lightblue', edgecolors='blue', linewidths=2, 
                  marker='s', zorder=3)
        # 绘制连接到走廊的线
        ax.plot([pos, pos], [corridor_y, corridor_y + room_offset], 
               'b--', linewidth=1.5, alpha=0.6, zorder=2)
        # 房间标签
        ax.text(pos, corridor_y + room_offset + 0.8, room.name, 
               fontsize=11, fontweight='bold', ha='center', 
               color='blue', va='bottom')
    
    # 右侧房间（在走廊下方）
    for room in right_rooms:
        pos = room.distance_from_exit1
        # 绘制房间标记
        ax.scatter(pos, corridor_y - room_offset, s=room_size*2, 
                  c='lightcoral', edgecolors='red', linewidths=2, 
                  marker='s', zorder=3)
        # 绘制连接到走廊的线
        ax.plot([pos, pos], [corridor_y, corridor_y - room_offset], 
               'r--', linewidth=1.5, alpha=0.6, zorder=2)
        # 房间标签
        ax.text(pos, corridor_y - room_offset - 0.8, room.name, 
               fontsize=11, fontweight='bold', ha='center', 
               color='red', va='top')
    
    # 人员1的路径（蓝色，实线）
    person1_color = '#2563eb'  # 更鲜明的蓝色
    person1_path_x = person1.path_positions
    person1_path_y = [corridor_y] * len(person1_path_x)
    
    # 绘制路径线（实线，加粗以突出显示）
    ax.plot(person1_path_x, person1_path_y, '-', color=person1_color, 
           linewidth=3.5, label='Person 1 Path', zorder=4, alpha=0.9, linestyle='-')
    
    # 绘制路径点和箭头
    for i in range(len(person1_path_x)):
        x = person1_path_x[i]
        label = person1.path_labels[i]
        
        if i == 0:
            # 起点（圆形标记）
            ax.scatter(x, corridor_y, s=130, c=person1_color, 
                      edgecolors='white', linewidths=2, 
                      marker='o', zorder=5)
            ax.text(x, corridor_y + 0.5, 'P1 Start', fontsize=9, 
                   fontweight='bold', color=person1_color, 
                   ha='center', va='bottom')
        else:
            # 房间位置（圆形标记）
            ax.scatter(x, corridor_y, s=100, c=person1_color, 
                      edgecolors='white', linewidths=1.5, 
                      marker='o', zorder=5)
            ax.text(x, corridor_y + 0.4, label, fontsize=10, 
                   fontweight='bold', color=person1_color, 
                   ha='center', va='bottom')
        
        # 绘制箭头（除了最后一个点）
        if i < len(person1_path_x) - 1:
            x_start = person1_path_x[i]
            x_end = person1_path_x[i + 1]
            dx = x_end - x_start
            
            if abs(dx) > 0.1:  # 只有当移动距离足够大时才绘制箭头
                # 箭头位置（路径的中间位置）
                arrow_x = (x_start + x_end) / 2
                arrow_y = corridor_y
                
                # 根据方向绘制箭头（加粗以突出显示）
                if dx > 0:
                    ax.annotate('', xy=(arrow_x + abs(dx)*0.15, arrow_y), 
                               xytext=(arrow_x - abs(dx)*0.15, arrow_y),
                               arrowprops=dict(arrowstyle='->', color=person1_color, 
                                             lw=2.5, alpha=0.95, shrinkA=0, shrinkB=0,
                                             headwidth=8, headlength=10), 
                               zorder=6)
                else:
                    ax.annotate('', xy=(arrow_x - abs(dx)*0.15, arrow_y), 
                               xytext=(arrow_x + abs(dx)*0.15, arrow_y),
                               arrowprops=dict(arrowstyle='->', color=person1_color, 
                                             lw=2.5, alpha=0.95, shrinkA=0, shrinkB=0,
                                             headwidth=8, headlength=10), 
                               zorder=6)
    
    # 人员2的路径（橙红色，虚线）
    person2_color = '#dc2626'  # 更鲜明的红色
    person2_path_x = person2.path_positions
    person2_path_y = [corridor_y] * len(person2_path_x)
    
    # 绘制路径线（虚线，加粗以突出显示）
    ax.plot(person2_path_x, person2_path_y, '--', color=person2_color, 
           linewidth=3.5, label='Person 2 Path', zorder=4, alpha=0.9, 
           dashes=(6, 3))
    
    # 绘制路径点和箭头
    for i in range(len(person2_path_x)):
        x = person2_path_x[i]
        label = person2.path_labels[i]
        
        if i == 0:
            # 起点（方形标记，与person1区分）
            ax.scatter(x, corridor_y, s=130, c=person2_color, 
                      edgecolors='white', linewidths=2, 
                      marker='s', zorder=5)
            ax.text(x, corridor_y - 0.5, 'P2 Start', fontsize=9, 
                   fontweight='bold', color=person2_color, 
                   ha='center', va='top')
        else:
            # 房间位置（方形标记）
            ax.scatter(x, corridor_y, s=100, c=person2_color, 
                      edgecolors='white', linewidths=1.5, 
                      marker='s', zorder=5)
            ax.text(x, corridor_y - 0.4, label, fontsize=10, 
                   fontweight='bold', color=person2_color, 
                   ha='center', va='top')
        
        # 绘制箭头（除了最后一个点）
        if i < len(person2_path_x) - 1:
            x_start = person2_path_x[i]
            x_end = person2_path_x[i + 1]
            dx = x_end - x_start
            
            if abs(dx) > 0.1:  # 只有当移动距离足够大时才绘制箭头
                # 箭头位置（路径的中间位置）
                arrow_x = (x_start + x_end) / 2
                arrow_y = corridor_y
                
                # 根据方向绘制箭头（加粗以突出显示）
                if dx > 0:
                    ax.annotate('', xy=(arrow_x + abs(dx)*0.15, arrow_y), 
                               xytext=(arrow_x - abs(dx)*0.15, arrow_y),
                               arrowprops=dict(arrowstyle='->', color=person2_color, 
                                             lw=2.5, alpha=0.95, shrinkA=0, shrinkB=0,
                                             headwidth=8, headlength=10), 
                               zorder=6)
                else:
                    ax.annotate('', xy=(arrow_x - abs(dx)*0.15, arrow_y), 
                               xytext=(arrow_x + abs(dx)*0.15, arrow_y),
                               arrowprops=dict(arrowstyle='->', color=person2_color, 
                                             lw=2.5, alpha=0.95, shrinkA=0, shrinkB=0,
                                             headwidth=8, headlength=10), 
                               zorder=6)
    
    # 设置坐标轴
    ax.set_xlim(-2, hallway_length + 2)
    ax.set_ylim(-5, 5)
    ax.set_xlabel('距离 Exit1 的位置 (米)', fontsize=12, fontweight='bold')
    ax.set_ylabel('位置', fontsize=12, fontweight='bold')
    ax.set_xticks(range(0, int(hallway_length) + 1, 5))
    ax.set_yticks([])
    
    # 添加标题
    ax.set_title('Room Inspection Strategy Visualization - Nearest Neighbor Greedy Algorithm', 
                fontsize=14, fontweight='bold', pad=20)
    
    # 添加图例
    ax.legend(loc='upper left', fontsize=11, framealpha=0.9)
    
    # 添加简化的统计信息（去掉框，只显示关键信息）
    person1_rooms_str = ', '.join(person1.rooms_assigned)
    person2_rooms_str = ', '.join(person2.rooms_assigned)
    
    stats_text = (
        f'Person 1: Rooms [{person1_rooms_str}] | Distance: {person1.total_distance:.1f}m | Time: {person1.total_time:.1f}s ({person1.total_time/60:.2f}min)\n'
        f'Person 2: Rooms [{person2_rooms_str}] | Distance: {person2.total_distance:.1f}m | Time: {person2.total_time:.1f}s ({person2.total_time/60:.2f}min)\n'
        f'Total Distance: {person1.total_distance + person2.total_distance:.1f}m | Max Time: {max(person1.total_time, person2.total_time):.1f}s | Time Diff: {abs(person1.total_time - person2.total_time):.1f}s'
    )
    
    ax.text(0.02, 0.98, stats_text, fontsize=9, 
           va='top', ha='left', transform=ax.transAxes,
           family='monospace', bbox=dict(boxstyle='round,pad=0.5', 
                                         facecolor='white', edgecolor='gray', 
                                         alpha=0.7, linewidth=0.5))
    
    # 调整布局
    plt.tight_layout()
    
    # 保存图片
    plt.savefig(save_path, dpi=300, bbox_inches='tight')
    print(f"图片已保存为 {save_path}")
    
    # 显示图片（如果环境支持）
    try:
        plt.show()
    except:
        print("注意：无法显示图形窗口，但图片已保存")
    finally:
        plt.close()


def main():
    """主函数"""
    # 创建算法实例并运行
    algorithm = RoomInspectionGreedy(hallway_length=30.0)
    
    # 运行算法
    person1, person2 = algorithm.greedy_assign_rooms(
        person1_start=0.0,      # 人员1从exit1开始
        person2_start=30.0,     # 人员2从exit2开始
        hallway_speed=1.5,      # 走廊移动速度 1.5米/秒
        room_speed=1.0,         # 房间内移动速度 1.0米/秒
        strategy="nearest"      # 使用最近邻策略
    )
    
    # 打印结果
    algorithm.print_results(person1, person2)
    
    # 可视化
    print("\n正在生成可视化图表...")
    visualize_strategy(person1, person2, algorithm)


if __name__ == "__main__":
    main()

File: visualize_strategy.m
% 房间检查策略可视化脚本
% 策略1: 最近邻贪心算法
% 
% 使用方法：
% 1. 直接运行此脚本（使用内置数据）
% 2. 或先运行 generate_matlab_data.py 生成数据文件，然后运行此脚本

clear; clc; close all;

%% 尝试加载数据文件（如果存在）
if exist('strategy_data.m', 'file') == 2
    fprintf('正在加载 strategy_data.m ...\n');
    run('strategy_data.m');
    fprintf('数据文件加载成功！\n\n');
else
    fprintf('未找到 strategy_data.m，使用默认数据...\n');
    fprintf('提示：运行 generate_matlab_data.py 可以生成最新的数据文件\n\n');
    
    %% 默认参数设置
    hallway_length = 30;  % 走廊长度（米）
    exit1_pos = 0;        % exit1位置
    exit2_pos = 30;       % exit2位置
    
    % 房间位置（距离exit1的距离，米）
    rooms = struct();
    rooms.L1 = 5;   % 左侧房间1
    rooms.L2 = 15;  % 左侧房间2
    rooms.L3 = 25;  % 左侧房间3
    rooms.R1 = 25;  % 右侧房间1（距离exit2 5m = 距离exit1 25m）
    rooms.R2 = 15;  % 右侧房间2（距离exit2 15m = 距离exit1 15m）
    rooms.R3 = 5;   % 右侧房间3（距离exit2 25m = 距离exit1 5m）
    
    % 人员1的路径（从算法结果）
    person1_path = [0, 5, 5, 15, 15];  % 起始位置 -> L1 -> R3 -> L2 -> R2
    person1_rooms = {'Start', 'L1', 'R3', 'L2', 'R2'};
    person1_total_distance = 15.00;
    person1_total_time = 19.00;
    
    % 人员2的路径（从算法结果）
    person2_path = [30, 25, 25];  % 起始位置 -> L3 -> R1
    person2_rooms = {'Start', 'L3', 'R1'};
    person2_total_distance = 5.00;
    person2_total_time = 7.00;
    
    % 统计信息
    total_distance = 20.00;
    max_completion_time = 19.00;
    time_difference = 12.00;
end

%% 创建图形
figure('Position', [100, 100, 1200, 800]);
hold on;
grid on;
axis equal;

%% 不绘制走廊背景，只添加出口标签
corridor_y = 0;
text(exit1_pos-1, corridor_y, 'Exit1', 'FontSize', 12, 'FontWeight', 'bold', ...
     'HorizontalAlignment', 'right');
text(exit2_pos+1, corridor_y, 'Exit2', 'FontSize', 12, 'FontWeight', 'bold', ...
     'HorizontalAlignment', 'left');

%% 绘制房间（左侧和右侧分开显示）
room_offset = 2;  % 房间距离走廊的偏移
room_size = 1.5;  % 房间标记大小

% 左侧房间（在走廊上方）
left_rooms_pos = [rooms.L1, rooms.L2, rooms.L3];
for i = 1:length(left_rooms_pos)
    pos = left_rooms_pos(i);
    room_name = sprintf('L%d', i);
    % 绘制房间标记
    plot(pos, corridor_y + room_offset, 's', 'MarkerSize', room_size*10, ...
         'MarkerFaceColor', [0.8, 0.9, 1.0], 'MarkerEdgeColor', 'b', 'LineWidth', 2);
    % 绘制连接到走廊的线
    plot([pos, pos], [corridor_y, corridor_y + room_offset], 'b--', 'LineWidth', 1);
    % 房间标签
    text(pos, corridor_y + room_offset + 0.8, room_name, 'FontSize', 11, ...
         'FontWeight', 'bold', 'HorizontalAlignment', 'center', 'Color', 'b');
end

% 右侧房间（在走廊下方）
right_rooms_pos = [rooms.R1, rooms.R2, rooms.R3];
right_room_names = {'R1', 'R2', 'R3'};
for i = 1:length(right_rooms_pos)
    pos = right_rooms_pos(i);
    room_name = right_room_names{i};
    % 绘制房间标记
    plot(pos, corridor_y - room_offset, 's', 'MarkerSize', room_size*10, ...
         'MarkerFaceColor', [1.0, 0.9, 0.8], 'MarkerEdgeColor', 'r', 'LineWidth', 2);
    % 绘制连接到走廊的线
    plot([pos, pos], [corridor_y, corridor_y - room_offset], 'r--', 'LineWidth', 1);
    % 房间标签
    text(pos, corridor_y - room_offset - 0.8, room_name, 'FontSize', 11, ...
         'FontWeight', 'bold', 'HorizontalAlignment', 'center', 'Color', 'r');
end

%% 绘制人员1的路径（蓝色）
person1_color = [0, 0.4470, 0.7410];  % 蓝色
person1_label_added = false;

% 绘制完整的路径线
plot(person1_path, ones(size(person1_path)) * corridor_y, '-', ...
     'Color', person1_color, 'LineWidth', 4, 'DisplayName', 'Person 1 Path');

% 绘制路径点和房间标记，并添加箭头
step_counter = 0;
for i = 1:length(person1_path)
    x = person1_path(i);
    if i == 1
        % 起点
        plot(x, corridor_y, 'o', 'MarkerSize', 14, 'MarkerFaceColor', person1_color, ...
             'MarkerEdgeColor', 'w', 'LineWidth', 2.5);
        text(x, corridor_y + 0.7, 'P1 Start', 'FontSize', 9, ...
             'FontWeight', 'bold', 'Color', person1_color, ...
             'HorizontalAlignment', 'center', 'BackgroundColor', 'w');
    else
        % 房间位置
        step_counter = step_counter + 1;
        if i <= length(person1_rooms)
            room_name = person1_rooms{i};
            if ~strcmp(room_name, 'Start')
                plot(x, corridor_y, 's', 'MarkerSize', 12, 'MarkerFaceColor', person1_color, ...
                     'MarkerEdgeColor', 'w', 'LineWidth', 2);
                % 添加房间名称和步骤编号
                text(x, corridor_y + 0.6, sprintf('%d: %s', step_counter, room_name), ...
                     'FontSize', 9, 'FontWeight', 'bold', 'Color', person1_color, ...
                     'HorizontalAlignment', 'center', 'BackgroundColor', 'w');
            end
        end
    end
    
    % 添加箭头（除了最后一个点）
    if i < length(person1_path)
        x_start = person1_path(i);
        x_end = person1_path(i + 1);
        dx = x_end - x_start;
        
        if abs(dx) > 0.1  % 只有当移动距离足够大时才绘制箭头
            % 箭头位置（路径的中间位置，稍微偏向起点）
            arrow_x = x_start + dx * 0.4;
            arrow_y = corridor_y;
            
            % 使用quiver绘制箭头（箭头长度为路径段长度的30%）
            arrow_length = dx * 0.3;
            quiver(arrow_x, arrow_y, arrow_length, 0, 'Color', person1_color, ...
                   'LineWidth', 2.5, 'MaxHeadSize', 1.0, 'AutoScale', 'off');
        end
    end
end

%% 绘制人员2的路径（红色）
person2_color = [0.8500, 0.3250, 0.0980];  % 橙红色
person2_label_added = false;

% 绘制完整的路径线
plot(person2_path, ones(size(person2_path)) * corridor_y, '-', ...
     'Color', person2_color, 'LineWidth', 4, 'DisplayName', 'Person 2 Path');

% 绘制路径点和房间标记，并添加箭头
step_counter = 0;
for i = 1:length(person2_path)
    x = person2_path(i);
    if i == 1
        % 起点
        plot(x, corridor_y, 'o', 'MarkerSize', 14, 'MarkerFaceColor', person2_color, ...
             'MarkerEdgeColor', 'w', 'LineWidth', 2.5);
        text(x, corridor_y - 0.7, 'P2 Start', 'FontSize', 9, ...
             'FontWeight', 'bold', 'Color', person2_color, ...
             'HorizontalAlignment', 'center', 'BackgroundColor', 'w');
    else
        % 房间位置
        step_counter = step_counter + 1;
        if i <= length(person2_rooms)
            room_name = person2_rooms{i};
            if ~strcmp(room_name, 'Start')
                plot(x, corridor_y, 's', 'MarkerSize', 12, 'MarkerFaceColor', person2_color, ...
                     'MarkerEdgeColor', 'w', 'LineWidth', 2);
                % 添加房间名称和步骤编号
                text(x, corridor_y - 0.6, sprintf('%d: %s', step_counter, room_name), ...
                     'FontSize', 9, 'FontWeight', 'bold', 'Color', person2_color, ...
                     'HorizontalAlignment', 'center', 'BackgroundColor', 'w');
            end
        end
    end
    
    % 添加箭头（除了最后一个点）
    if i < length(person2_path)
        x_start = person2_path(i);
        x_end = person2_path(i + 1);
        dx = x_end - x_start;
        
        if abs(dx) > 0.1  % 只有当移动距离足够大时才绘制箭头
            % 箭头位置（路径的中间位置，稍微偏向起点）
            arrow_x = x_start + dx * 0.4;
            arrow_y = corridor_y;
            
            % 使用quiver绘制箭头（箭头长度为路径段长度的30%）
            arrow_length = dx * 0.3;
            quiver(arrow_x, arrow_y, arrow_length, 0, 'Color', person2_color, ...
                   'LineWidth', 2.5, 'MaxHeadSize', 1.0, 'AutoScale', 'off');
        end
    end
end

%% 添加图例和标题
legend({'Person 1 Path', 'Person 2 Path'}, 'Location', 'northwest', 'FontSize', 12);
title('房间检查策略可视化 - 最近邻贪心算法（策略1）', 'FontSize', 14, 'FontWeight', 'bold');
xlabel('距离 Exit1 的位置 (米)', 'FontSize', 12);
ylabel('位置', 'FontSize', 12);

% 设置坐标轴范围
xlim([-2, 32]);
ylim([-4, 4]);

% 添加网格
set(gca, 'XTick', 0:5:30);
set(gca, 'YTick', []);

%% 添加统计信息文本框
% 构建人员1的房间列表字符串
person1_rooms_str = strjoin(person1_rooms(2:end), ', ');
% 构建人员2的房间列表字符串
person2_rooms_str = strjoin(person2_rooms(2:end), ', ');

stats_text = {
    sprintf('人员1: 检查房间 [%s]', person1_rooms_str);
    sprintf('       总移动距离: %.2f 米', person1_total_distance);
    sprintf('       总时间: %.2f 分钟', person1_total_time);
    sprintf('');
    sprintf('人员2: 检查房间 [%s]', person2_rooms_str);
    sprintf('       总移动距离: %.2f 米', person2_total_distance);
    sprintf('       总时间: %.2f 分钟', person2_total_time);
    sprintf('');
    sprintf('总体: 总移动距离 %.2f 米', total_distance);
    sprintf('      最大完成时间: %.2f 分钟', max_completion_time);
    sprintf('      时间差: %.2f 分钟', time_difference);
};

text(32, 3, stats_text, 'FontSize', 10, 'VerticalAlignment', 'top', ...
     'HorizontalAlignment', 'left', 'BackgroundColor', 'w', ...
     'EdgeColor', 'k', 'Margin', 5);

%% 保存图片
print('room_inspection_strategy.png', '-dpng', '-r300');
fprintf('图片已保存为 room_inspection_strategy.png\n');

